scripts/generators/typescript.js

```diff
@@ -0,0 +1,155 @@
+"use strict";
+
+const t = require("../../packages/babel-types");
+const utils = require("./utils");
+
+const NODE_PREFIX = "";
+
+let code = `// NOTE: This file is autogenerated. Do not modify.
+// See scripts/generators/typescript.js for script used.
+
+export class ${NODE_PREFIX}Comment {
+  value: string;
+  start: number;
+  end: number;
+  loc: ${NODE_PREFIX}SourceLocation;
+}
+
+export class ${NODE_PREFIX}BlockComment extends ${NODE_PREFIX}Comment {
+  type: "BlockComment";
+}
+
+export class ${NODE_PREFIX}LineComment extends ${NODE_PREFIX}Comment {
+  type: "LineComment";
+}
+
+export class ${NODE_PREFIX}SourceLocation {
+  start: {
+    line: number;
+    column: number;
+  };
+
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+export class Node {
+  leadingComments: Array<${NODE_PREFIX}Comment> | null;
+  innerComments: Array<${NODE_PREFIX}Comment> | null;
+  trailingComments: Array<${NODE_PREFIX}Comment> | null;
+  start: number | null;
+  end: number | null;
+  loc: ${NODE_PREFIX}SourceLocation | null;
+}\n\n`;
+
+//
+
+const lines = [];
+
+for (const type in t.NODE_FIELDS) {
+  const fields = t.NODE_FIELDS[type];
+
+  const struct = ['type: "' + type + '";'];
+  const args = [];
+
+  Object.keys(t.NODE_FIELDS[type])
+    .sort((fieldA, fieldB) => {
+      const indexA = t.BUILDER_KEYS[type].indexOf(fieldA);
+      const indexB = t.BUILDER_KEYS[type].indexOf(fieldB);
+      if (indexA === indexB) return fieldA < fieldB ? -1 : 1;
+      if (indexA === -1) return 1;
+      if (indexB === -1) return -1;
+      return indexA - indexB;
+    })
+    .forEach(fieldName => {
+      const field = fields[fieldName];
+      const typeAnnotation = utils.stringifyValidator(
+        field.validate,
+        NODE_PREFIX
+      );
+
+      args.push(
+        t.toBindingIdentifierName(fieldName) +
+          (isNullable(field) ? "?: " : ": ") +
```

Comment:
To be sure, does babel make absolutely no distinction between `null` and `undefined`, or is `null` distinct? `FunctionDeclaration` `id`s come to mind.

Reply:
The only distinction I can find is when a field has a default value, `null` doesn't get parsed into `undefined` and instead throws an error. Updated.

-------

Here's the result of constructing a few types:

- If field is `optional`:
  - `undefined` Ok, becomes `null`
  - `null` Ok
- If field has a `default`:
  - `undefined` Ok, uses default values
  - `null` Error
- Otherwise, if field is required:
  - `undefined` Error
  - `null` Error